offsite-activity-signup/src/Code.js

```diff
@@ -0,0 +1,458 @@
+  var attendees = rows.map(function(row) {
+    var _ = row.shift(); // Ignore timestamp
+    var email = row.shift();
+    var autoAssign = row.pop();
+    // Find ranked items in the response data.
+    var preferences = row.reduce(function(prefs, value, index) {
+      var match = value.match(/(\d+).*/);
+      if (!match) {
+        return prefs;
+      }
+      var rank = parseInt(match[1]) - 1; // Convert ordinal to array index
+      prefs[rank] = index;
+      return prefs;
+    }, []);
+    if (autoAssign == 'Yes') {
+      // If auto assigning additional activites, append a randomized
+      // list of all the activities. These will then be considered
+      // as if the attendee ranked them.
+      var additionalChoices = shuffleArray_(allActivityIds);
+      preferences = preferences.concat(additionalChoices);
+    }
+    return {
+      email: email,
+      preferences: preferences,
+      assigned: [],
+    };
+  });
+  return attendees;
+}
+
+/**
+ * Simulates a large number of users responding to the form. This enables users
+ * to quickly experience the full solution without having to collect sufficient
+ * form responses through other means.
+ */
+function generateTestData_() {
+  var ss = SpreadsheetApp.getActiveSpreadsheet();
+  var sheet = findResponseSheetForForm_(ss);
+  if (!sheet) {
+    var msg = 'No response sheet found. Create the form and try again.';
+    SpreadsheetApp.getUi().alert(msg);
+  }
+  if (sheet.getLastRow() > 1) {
+    var msg = 'Response sheet is not empty, can not generate test data. ' +
+      'Remove responses and try again.';
+    SpreadsheetApp.getUi().alert(msg);
+    return;
+  }
+
+  var activities = loadActivitySchedule_(ss);
+  var choices = fillArray_([], activities.length, '');
+  range_(1, 5).forEach(function(value) {
+    choices[value] = toOrdinal_(value);
+  });
+
+  var rows = range_(1, NUM_TEST_USERS).map(function(value) {
+    var randomizedChoices = shuffleArray_(choices);
+    var email = Utilities.formatString('person%d@example.com', value);
+    return [new Date(), email].concat(randomizedChoices).concat(['Yes']);
+  });
+  bulkAppendRows_(sheet, rows);
+}
+
+/**
+ * Retrieve a sheet by name, creating it if it doesn't yet exist.
+ *
+ * @param {Spreadsheet} ss - Containing spreadsheet
+ * @Param {string} name - Name of sheet to return
+ * @return {Sheet} Sheet instance
+ */
+function findOrCreateSheetByName_(ss, name) {
+  var sheet = ss.getSheetByName(name);
+  if (sheet) {
+    return sheet;
+  }
+  return ss.insertSheet(name);
+}
+
+/**
+ * Faster version of appending multiple rows via ranges.
+ *
+ * @param {Sheet} sheet - Sheet to append to
+ * @param {object[][]} rows - Rows to append
+ */
+function bulkAppendRows_(sheet, rows) {
+  var startRow = sheet.getLastRow() + 1;
+  var startColumn = 1;
+  var numRows = rows.length;
+  var numColumns = rows[0].length; // NOTE: Assumes all rows equal length
```

Comment:
I believe setValues() fails if the rows aren't of equal length, so it should be a safe assumption.

Reply:
Indeed it does. Moved the comment up to the method description.